homeassistant/components/media_player/dlna_dmr.py

```diff
@@ -0,0 +1,717 @@
+
+
+class UpnpNotifyView(HomeAssistantView):
+    """Callback view for UPnP NOTIFY messages."""
+
+    url = '/api/dlna_dmr.notify'
+    name = 'api:dlna_dmr:notify'
+    requires_auth = False
+
+    def __init__(self, hass):
+        """Initializer."""
+        self.hass = hass
+        self._registered_services = {}
+        self._backlog = {}
+
+    def register(self, router):
+        """Register the view with a router."""
+        handler = request_handler_factory(self, self.async_notify)
+        router.add_route('notify', UpnpNotifyView.url, handler)
+
+    async def async_notify(self, request):
+        """Callback method for NOTIFY requests."""
+        if 'SID' not in request.headers:
+            return aiohttp.web.Response(status=412)
+
+        headers = request.headers
+        sid = headers['SID']
+        body = await request.text()
+
+        # find UpnpService by SID
+        if sid not in self._registered_services:
+            _LOGGER.debug('Storing NOTIFY in backlog for SID: %s', sid)
+            self._backlog[sid] = {'headers': headers, 'body': body}
+            return aiohttp.web.Response(status=202)
+
+        service = self._registered_services[sid]
+        status = service.on_notify(headers, body)
+        return aiohttp.web.Response(status=status)
+
+    @property
+    def callback_url(self):
+        """Full URL to be called by device/service."""
+        base_url = self.hass.config.api.base_url
+        return urllib.parse.urljoin(base_url, self.url)
+
+    def register_service(self, sid, service):
+        """Register a UpnpService under SID."""
+        if sid in self._registered_services:
+            raise RuntimeError('SID {} already registered.'.format(sid))
+
+        self._registered_services[sid] = service
+
+        if sid in self._backlog:
+            item = self._backlog[sid]
+            _LOGGER.debug('Re-playing backlogged NOTIFY for SID: %s', sid)
+            service.on_notify(item['headers'], item['body'])
+            del self._backlog[sid]
+
+    def unregister_service(self, sid):
+        """Unregister service by SID."""
+        if sid in self._registered_services:
+            del self._registered_services[sid]
+
+
+class HassUpnpRequester(object):
+    """async_upnp_client.UpnpRequester for home-assistant."""
+
+    def __init__(self, hass):
+        """Initializer."""
+        self.hass = hass
+
+    async def async_http_request(self, method, url, headers=None, body=None):
+        """Do a HTTP request."""
+        session = async_get_clientsession(self.hass)
+        with async_timeout.timeout(5, loop=self.hass.loop):
+            response = await session.request(method,
+                                             url,
+                                             headers=headers,
+                                             data=body)
+            response_body = await response.text()
+            await response.release()
+        await asyncio.sleep(0.25)
```

Comment:
Unnecessary sleep?

Reply:
Might be a duplicate of another sleep. I'll remove it and test it.

Reply:
Now I remember... I was having problems like these:
ConnectionResetError: Cannot write to closing transport
aiohttp.client_exceptions.ClientOSError: [Errno None] Can not write request body for http://192.168.178.10:49152/ctl/AVTransport

Any clue on how to solve these?

Reply:
The actual tracebacks:
```
Traceback (most recent call last):
  File "/home/steven/.virtualenvs/home-assistant/lib/python3.6/site-packages/aiohttp/client_reqrep.py", line 474, in write_bytes
    await self.body.write(writer)
  File "/home/steven/.virtualenvs/home-assistant/lib/python3.6/site-packages/aiohttp/payload.py", line 181, in write
    await writer.write(self._value)
  File "/home/steven/.virtualenvs/home-assistant/lib/python3.6/site-packages/aiohttp/http_writer.py", line 90, in write
    self._write(chunk)
  File "/home/steven/.virtualenvs/home-assistant/lib/python3.6/site-packages/aiohttp/http_writer.py", line 57, in _write
    raise ConnectionResetError('Cannot write to closing transport')
ConnectionResetError: Cannot write to closing transport

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/steven/src/ha/home-assistant/homeassistant/core.py", line 1007, in _event_to_service_call
    await service_handler.func(service_call)
  File "/home/steven/src/ha/home-assistant/homeassistant/components/media_player/__init__.py", line 441, in async_service_handler
    await getattr(player, method['method'])(**params)
  File "/home/steven/src/ha/home-assistant/homeassistant/components/media_player/dlna_dmr.py", line 588, in async_play_media
    await self.async_media_play()
  File "/home/steven/src/ha/home-assistant/homeassistant/components/media_player/dlna_dmr.py", line 544, in async_media_play
    await action.async_call(InstanceID=0, Speed='1')
  File "/home/steven/src/ha/async_upnp_client/async_upnp_client/async_upnp_client.py", line 438, in async_call
    await self.service.requester.async_http_request('POST', url, headers, body)
  File "/home/steven/src/ha/home-assistant/homeassistant/components/media_player/dlna_dmr.py", line 325, in async_http_request
    data=body)
  File "/home/steven/.virtualenvs/home-assistant/lib/python3.6/site-packages/aiohttp/client.py", line 387, in _request
    await resp.start(conn)
  File "/home/steven/.virtualenvs/home-assistant/lib/python3.6/site-packages/aiohttp/client_reqrep.py", line 748, in start
    message, payload = await self._protocol.read()
  File "/home/steven/.virtualenvs/home-assistant/lib/python3.6/site-packages/aiohttp/streams.py", line 533, in read
    await self._waiter
aiohttp.client_exceptions.ClientOSError: [Errno None] Can not write request body for http://192.168.178.10:49152/ctl/AVTransport
```